src/settings/types/Duration.h

```diff
@@ -0,0 +1,70 @@
+//Copyright (c) 2018 Ultimaker B.V.
+//CuraEngine is released under the terms of the AGPLv3 or higher.
+
+#ifndef DURATION_H
+#define DURATION_H
+
+namespace cura
+{
+
+/*
+ * \brief Represents a duration in seconds.
+ *
+ * This is a facade. It behaves like a double, only it can't be negative.
+ */
+struct Duration
```

Comment:
If we'd like to inline these functions completely transparent, we should not use inheritance. A virtual table would destroy performance. If it goes out of hand with the overloads we should look into the templated one.

Reply:
Inheritance introduces vtables, right? templates only duplicate implementation and affect compile time performance...

Reply:
Without inheritance, all of these classes (AngleDegrees, AngleRadians, Temperature, LayerIndex, etc) would become one class. How will that class be named then? `Number`? And everything that stores a layer index throughout our code base has to declare `Number<AngleRadians> support_infill_angle;`. That also doesn't look very nice.

Reply:
Uhm yeah templates without inheritance is a bit nasty. I didn't really think about that yet.

Still there's a lot of code duplication, so it might be worthwhile to think about these classes.

Also there is some discrepancy between the classes which I don't see any reason for.
Why does Velocity have all these operators, while Temperature has none?

> That also doesn't look very nice.

you can always solve stuff like that with `using Temperature = NumberWrapper<...>`

I do like to have a separate class for the Angles, because of their modulus arithmetic. However, you can make one templated class out of it with a number as template argument. `template<double> struct ModuloNumber` with `using AngleDegrees = ModuloNumber<360>` and `using AngleDegrees = ModuloNumber<2.0 * M_PI>`.